### PR TITLE
Postgres Re-write

### PR DESCRIPTION
**Pull request by binaryoverload**
# Pull request

 - [x] I have read and agreed to the [code of conduct](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CODE_OF_CONDUCT.md).
 - [x] I have read and complied with the [contributing guidelines](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CONTRIBUTING.md).
 - [x] What I'm implementing was assigned to me or was not being worked on and is a somewhat requested feature. For reference see our [GitHub projects](https://github.com/orgs/CascadeBot/projects).
 - [ ] I have tested all of my changes.
 
## Added/Changed
 - Creation of schema for postgres
 - All entities for the rewrite

Closing #264

### CODE REVIEW

**Comment by binaryoverload** [comment]
jenkins build

File: src/main/java/org/cascadebot/cascadebot/CascadeBot.java [hunk 1]
```diff
@@ -171,6 +172,8 @@ private void init() {
             );
         }
 
+        PostgresManager postgresManager = new PostgresManager("localhost", "postgres", "postgres", "jvs123", "");
```
Comment (weeryan17) [inline src/main/java/org/cascadebot/cascadebot/CascadeBot.java]: Nice password

File: src/main/kotlin/org/cascadebot/cascadebot/data/PostgresManager.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import org.hibernate.SessionFactory
+import org.hibernate.cfg.Configuration
+import org.reflections.Reflections
+import java.net.URLEncoder
+import java.util.Properties
+import java.util.function.Consumer
+import javax.persistence.Entity
+import org.cascadebot.cascadebot.data.transaction as kotlinTransaction
+
+class PostgresManager(hosts: String, database: String, username: String, password: String, val options: String) {
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/data/PostgresManager.kt]: Would options be better as some sort of map? As far as I'm aware options are query parameters which are key value.

File: src/main/kotlin/org/cascadebot/cascadebot/data/PostgresManager.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        require(database.isNotBlank()) { "database cannot be empty or blank!" }
+        require(username.isNotBlank()) { "username cannot be empty or blank!" }
+        require(password.isNotBlank()) { "password cannot be empty or blank!" }
+
+        val dbConfig = Configuration()
+
+        val reflections = Reflections("org.cascadebot.cascadebot.data.entities")
+        val classes: Set<Class<*>> = reflections.getTypesAnnotatedWith(Entity::class.java)
+
+        classes.forEach { dbConfig.addAnnotatedClass(it) }
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/data/PostgresManager.kt]: This should probably be added/incorporated into our reflection utils

File: src/main/kotlin/org/cascadebot/cascadebot/data/entities/GuildAutoRoleEntity.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import javax.persistence.Entity
+import javax.persistence.Id
+import javax.persistence.Table
+
+@Entity
+@Table(name = "guild_autorole")
+class GuildAutoRoleEntity(guildId: Long, roleId: Long) {
+
+    @Id
+    @Column(name = "id", nullable = false)
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/data/entities/GuildAutoRoleEntity.kt]: Why are we setting nullable to false?

File: src/main/kotlin/org/cascadebot/cascadebot/data/EnumDBType.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package org.cascadebot.cascadebot.data
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor
+import org.hibernate.type.EnumType
+import java.sql.PreparedStatement
+import java.sql.Types
+
+class EnumDBType : EnumType<Enum<*>>() {
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/data/EnumDBType.kt:13]: Out of curiosity, why does this need to exist?

File: src/main/kotlin/org/cascadebot/cascadebot/data/entities/package-info.java [hunk 6]
```diff
@@ -0,0 +1,10 @@
+@TypeDefs({
+        @TypeDef(name = "psql-enum", typeClass = EnumDBType.class),
+        @TypeDef(name = "list-array", typeClass = ListArrayType.class)
+})
+package org.cascadebot.cascadebot.data.entities;
+
+import com.vladmihalcea.hibernate.type.array.ListArrayType;
+import org.cascadebot.cascadebot.data.EnumDBType;
+import org.hibernate.annotations.TypeDef;
+import org.hibernate.annotations.TypeDefs;
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/data/entities/package-info.java]: What is this file and why does it exist

**Review by weeryan17 — CHANGES_REQUESTED** [review]
Hey binary, while we're going in and doing all these data changes, can we also move all of guild data stuff to their own package. It's going to happen at some time anyways for donation stuffs as I was going to split user data stuff into it's own package so I figure it might as well happen now.

**Pull request closed without merging**